### PR TITLE
Add resource manager canCreate connection for the tcp proxy

### PR DESCRIPTION
**Pull request by junr03**

### CODE REVIEW

File: source/common/filter/tcp_proxy.cc [hunk 1]
```diff
@@ -41,13 +46,22 @@ TcpProxyStats TcpProxyConfig::generateStats(const std::string& name, Stats::Stor
...
+  Upstream::ResourceManager& cluster_resource_manager =
+      cluster_manager_.get(config_->clusterName())
+          ->resourceManager(Upstream::ResourcePriority::Default);
+  if (cluster_resource_manager.connections().canCreate()) {
+    Upstream::Host::CreateConnectionData conn_info =
+        cluster_manager_.tcpConnForCluster(config_->clusterName());
+
+    upstream_connection_ = std::move(conn_info.connection_);
+    read_callbacks_->upstreamHost(conn_info.host_description_);
+    cluster_resource_manager.connections().inc();
```
Comment (mattklein123) [inline source/common/filter/tcp_proxy.cc]: do inc() in else block, then you don't need dec() below

**Comment by mattklein123** [comment]
Please make sure tcp_proxy docs discuss use of resource manager and have the appropriate cross links

File: docs/intro/arch_overview/tcp_proxy.rst [hunk 2]
```diff
@@ -9,4 +9,9 @@ clusters. It can be used by itself as an stunnel replacement, or in conjunction
 such as the :ref:`MongoDB filter <arch_overview_mongo>` or the :ref:`rate limit
 <config_network_filters_rate_limit>` filter.
 
+The TCP proxy filter will respect connection limits
```
Comment (mattklein123) [inline docs/intro/arch_overview/tcp_proxy.rst]: can you link to circuit breaking configuration page

File: source/common/filter/tcp_proxy.cc [hunk 3]
```diff
@@ -41,13 +46,21 @@ TcpProxyStats TcpProxyConfig::generateStats(const std::string& name, Stats::Stor
...
+        cluster_manager_.tcpConnForCluster(config_->clusterName());
+
+    upstream_connection_ = std::move(conn_info.connection_);
+    read_callbacks_->upstreamHost(conn_info.host_description_);
+    if (!upstream_connection_) {
+      read_callbacks_->connection().close(Network::ConnectionCloseType::NoFlush);
+      return;
+    }
+    cluster_resource_manager.connections().inc();
+  } else {
```
Comment (mattklein123) [inline source/common/filter/tcp_proxy.cc]: delete redundant else case

File: source/common/filter/tcp_proxy.cc [hunk 4]
```diff
@@ -41,13 +46,21 @@ TcpProxyStats TcpProxyConfig::generateStats(const std::string& name, Stats::Stor
...
+          ->resourceManager(Upstream::ResourcePriority::Default);
+  if (cluster_resource_manager.connections().canCreate()) {
+    Upstream::Host::CreateConnectionData conn_info =
+        cluster_manager_.tcpConnForCluster(config_->clusterName());
+
+    upstream_connection_ = std::move(conn_info.connection_);
+    read_callbacks_->upstreamHost(conn_info.host_description_);
+    if (!upstream_connection_) {
+      read_callbacks_->connection().close(Network::ConnectionCloseType::NoFlush);
+      return;
```
Comment (mattklein123) [inline source/common/filter/tcp_proxy.cc]: remove return

File: source/common/filter/tcp_proxy.cc [hunk 5]
```diff
@@ -41,13 +46,21 @@ TcpProxyStats TcpProxyConfig::generateStats(const std::string& name, Stats::Stor
...
+    }
```
Comment (mattklein123) [inline source/common/filter/tcp_proxy.cc]: else

File: docs/intro/arch_overview/tcp_proxy.rst (hunk 2 above)
Comment (junr03) [inline docs/intro/arch_overview/tcp_proxy.rst]: added

File: source/common/filter/tcp_proxy.cc (hunk 3 above)
Comment (junr03) [inline source/common/filter/tcp_proxy.cc]: cleaner version in latest commit.

File: source/common/filter/tcp_proxy.cc (hunk 4 above)
Comment (junr03) [inline source/common/filter/tcp_proxy.cc]: cleaner version in latest commit.

File: source/common/filter/tcp_proxy.cc (hunk 5 above)
Comment (junr03) [inline source/common/filter/tcp_proxy.cc]: cleaner version in latest commit.

File: source/common/filter/tcp_proxy.cc [hunk 6]
```diff
@@ -41,6 +46,13 @@ TcpProxyStats TcpProxyConfig::generateStats(const std::string& name, Stats::Stor
 }
 
 void TcpProxy::initializeUpstreamConnection() {
+  Upstream::ResourceManager& cluster_resource_manager =
+      cluster_manager_.get(config_->clusterName())
+          ->resourceManager(Upstream::ResourcePriority::Default);
+
+  if (!cluster_resource_manager.connections().canCreate()) {
+    return;
```
Comment (mattklein123) [inline source/common/filter/tcp_proxy.cc]: This is not correct. You need to close the downstream connection (and test for that).

Comment (junr03) [inline source/common/filter/tcp_proxy.cc]: I am closing downstream, and testing for that in the UpstreamConnectionLimit test.

**Pull request merged**